### DISCUSSION TITLE
Allow required user fields to be empty initially

### DISCUSSION THREAD

**Discussion by jimafisk** [pocketbase/pocketbase#886] [Q&A] [answered]
I'm testing [v0.8.0-rc1](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc1) and was curious if there was a way to require that certain user fields have values, but also let them be blank/null on the initial user creation.

The desired workflow:

1. Users create an account with email / password
2. They verify their email addresses
3. Then fill out their required info: `first_name`, `last_name`, etc.
4. Allow user to later edit `first_name`, `last_name`, etc, but never allow it to be completely removed (blank submission)

This was possible in `v0.7.x` releases (https://github.com/pocketbase/pocketbase/discussions/225#discussioncomment-3511354):

> empty profile will be created automatically in the background during the user creation

I could submit placeholder values for these fields on user creation, but I was wondering if there was an official way you expected folks to handle this scenario. Thank you!

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure what would be the best way to handle this.

The old behavior was a limitation of the previous implementation, where the current one is consistent with the regular collection fields handling.

If a field is marked "required" then in my opinion it must be always nonempty.  Adding a separate option to control the required behavior for create and update may complicate and clutter the UI and I'm not sure if it is a good idea because not everyone will want this.

For now, if you want the previous behavior, you could:

- Apply the update-only required validations client-side (this obviously is acceptable only for non-sensitive/meta fields)

- Remove the required constraint from the fields and manually check them against an empty string in the "Update API rule", eg. `@request.data.first_name != "" && @request.data.last_name != ""`

- Create a separate `profiles` base collection and associate it with an optional `relation` field in your `users` collection (this is the closest you'll get to the previous behavior)

- Handle the validation via the Go `OnRecordBefore*` event hooks

**Reply by jimafisk** [reply]
Thanks for outlining these 4 options! I like the second one (removing the required constraint and adding Update Rule checks), the only downside is that I would assume the errors returned would be less descriptive (`first_name: Invalid value format` vs `first_name: cannot be blank`).

I tried testing this out, but running an update that doesn't pass the API Rules simply results in a 404 currently:

<details>
<summary>Client error</summary>

```
ClientResponseError 404: The requested resource wasn't found.
    at new o (pocketbase.es.mjs:1:2078)
    at pocketbase.es.mjs:1:31752
    at async HTMLButtonElement.setProfileInfo (profile.js:659:4)
```

</details>

<details>
<summary>PocketBase logs</summary>

```
{
  "errorDetails": "sql: no rows in result set",
  "errorMessage": "The requested resource wasn't found."
}
```

</details>

(running an update that passes the API Rules does patch the record correctly)

I think this is probably related to the error wrapping we discussed on another thread: https://github.com/pocketbase/pocketbase/discussions/196#discussioncomment-4022174

**Reply by ganigeorgiev** [reply]
@jimafisk Ah, this is expected, I actually forgot to mention it but you'll not get individual validation errors in case of mismatched filter. It is either the record match the entire API rule or not, so this option may not be suitable if you rely on the server-side field error messages.

**Reply by jimafisk** [reply]
I know you have higher priorities, but it _might_ be cool in the future to allow setting individual error messages for specific conditions in an API Rule. You'd need a way to combine conditions with OR and AND, so the UI could get a little cumbersome.

**Comment by jimafisk** [comment]
@ganigeorgiev this might be a crazy idea, but what about providing the option to expand a collection within another collection in the admin interface? Visually it would look similar to how you previously managed profiles within users. Essentially you'd have a toggle for relation fields that would allow the fields from collection B to appear inline with the record from collection A. I'd still want collection B to live separately, but it would be nice if I could easily make that association from collection A and even make edits to the record (not the collection) right there as well. I realize this could be a nightmare to implement and probably isn't anywhere near the roadmap, just wanted to throw it out there! Thanks!

**Reply by ganigeorgiev** [reply]
It is not entirely the same as what you describe, but I have something similar in my todo to improve the relation association UI.
I've already started experimenting with that in v0.8.0 by adding controls to create a record from the relation dropdown and directly associate it on save, but inline edit/delete and fields expansion in the list for now is left for future implementation since I'm not entirely sure how this could be done without cluttering the UI and making the collections management too overwhelming and I don't want to rush it (_that's why I also left https://github.com/pocketbase/pocketbase/issues/881 unassigned for now_).
